File: server-springboot/src/main/java/com/campusbazaar/model/Listing.java
package com.campusbazaar.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.CompoundIndexes;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.index.TextIndexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document("listings")
@CompoundIndexes({
        @CompoundIndex(name = "category_status_created", def = "{ 'category': 1, 'status': 1, 'createdAt': -1 }")
})
public class Listing {

    public enum Category { Books, Electronics, Clothing, Furniture, Services, Other }
    public enum Condition { New, Good, Fair }
    public enum Status { active, sold, expired }

    @Id
    private String id;

    @TextIndexed
    private String title;

    @TextIndexed
    private String description;

    private double price;

    @Builder.Default
    private List<String> images = new ArrayList<>();

    private Category category;
    private Condition condition;

    @Indexed
    private String seller;       // User id

    @Builder.Default
    @Indexed
    private Status status = Status.active;

    @Builder.Default
    private long views = 0;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;
}

File: server/models/Listing.js
const mongoose = require('mongoose');

const CATEGORIES = ['Books', 'Electronics', 'Clothing', 'Furniture', 'Services', 'Other'];
const CONDITIONS = ['New', 'Good', 'Fair'];

const listingSchema = new mongoose.Schema(
  {
    title: { type: String, required: true, trim: true, maxlength: 120 },
    description: { type: String, required: true, maxlength: 4000 },
    price: { type: Number, required: true, min: 0 },
    images: {
      type: [String],
      validate: [(arr) => arr.length <= 5, 'Maximum 5 images allowed']
    },
    category: { type: String, enum: CATEGORIES, required: true },
    condition: { type: String, enum: CONDITIONS, required: true },
    seller: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    status: { type: String, enum: ['active', 'sold', 'expired'], default: 'active', index: true },
    views: { type: Number, default: 0 }
  },
  { timestamps: true }
);

listingSchema.index({ title: 'text', description: 'text' });
listingSchema.index({ category: 1, status: 1, createdAt: -1 });

module.exports = mongoose.model('Listing', listingSchema);
module.exports.CATEGORIES = CATEGORIES;
module.exports.CONDITIONS = CONDITIONS;
